Content.Server/Connection/ConnectionManager.cs

```diff
@@ -175,26 +207,191 @@ private async Task NetMgrOnConnecting(NetConnectingArgs e)
                 {
-                    var msg = Loc.GetString(_cfg.GetCVar(CCVars.WhitelistReason));
-                    // was the whitelist playercount changed?
-                    if (min > 0 || max < int.MaxValue)
-                        msg += "\n" + Loc.GetString("whitelist-playercount-invalid", ("min", min), ("max", max));
-                    return (ConnectionDenyReason.Whitelist, msg, null);
+                    if (!IsValid(whitelist, _plyMgr.PlayerCount))
+                    {
+                        // Not valid for current player count.
+                        continue;
+                    }
+
+                    var whitelistStatus = await IsWhitelisted(whitelist, e.UserData, _sawmill);
+                    if (!whitelistStatus.isWhitelisted)
+                    {
+                        // Not whitelisted.
+                        return (ConnectionDenyReason.Whitelist, Loc.GetString("whitelist-fail-prefix", ("msg", whitelistStatus.denyMessage!)), null);
+                    }
+
+                    // Whitelisted, don't check any more.
+                    break;
                 }
             }
 
             return null;
         }
 
+        public bool IsValid(PlayerConnectionWhitelistPrototype whitelist, int playerCount)
+        {
+            return playerCount >= whitelist.MinimumPlayers && playerCount <= whitelist.MaximumPlayers;
+        }
+
+        public async Task<(bool isWhitelisted, string? denyMessage)> IsWhitelisted(PlayerConnectionWhitelistPrototype whitelist, NetUserData data, ISawmill sawmill)
+        {
+            foreach (var condition in whitelist.Conditions)
+            {
+                var matched = false;
+                string denyMessage;
+                switch (condition.GetType())
+                {
+                    case { } t when t == typeof(ConditionAlwaysMatch):
+                        matched = true;
+                        denyMessage = Loc.GetString("whitelist-always-deny");
+                        break;
+                    case { } t when t == typeof(ConditionManualWhitelist):
+                        matched = !(await _db.GetWhitelistStatusAsync(data.UserId));
+                        denyMessage = Loc.GetString("whitelist-manual");
+                        break;
+                    case { } t when t == typeof(ConditionManualBlacklist):
+                        var blacklisted = await _db.GetBlacklistStatusAsync(data.UserId);
+                        matched = blacklisted;
+                        denyMessage = Loc.GetString("whitelist-blacklisted");
+                        break;
+                    case { } t when t == typeof(ConditionNotesDateRange):
+                        var conditionNotes = (ConditionNotesDateRange)condition;
+                        var remarks = await GetAdminRemarks(data.UserId);
+                        remarks = remarks.Where(x => x.CreatedAt > DateTime.Now.AddDays(-conditionNotes.Range)).ToList();
```

Comment:
This isn't a major deal to me, but it's always poor form to run `DateTime.Now` in a loop like this as you'll have to re-fetch the system time every loop. And subtract the date range too.

Also do this with UTC instead. I'm not sure it matters but it cannot hurt to be sure.